gax/src/test/java/com/google/api/gax/batching/ThresholdBatcherTest.java

```diff
@@ -299,11 +309,52 @@ public void testBatchingFlowControlExceptionRecovery() throws Exception {
     for (SimpleBatch batch : receiver.getBatches()) {
       actual.add(batch.getIntegers());
     }
-    Truth.assertThat(actual).isEqualTo(expected);
+    assertThat(actual).isEqualTo(expected);
+
+    assertThat(trackedFlowController.getElementsReserved())
+        .isEqualTo(trackedFlowController.getElementsReleased());
+    assertThat(trackedFlowController.getBytesReserved())
+        .isEqualTo(trackedFlowController.getBytesReleased());
+  }
+
+  @Test
+  public void testBatchingFailedRPC() throws Exception {
+    AccumulatingBatchReceiver<SimpleBatch> receiver =
+        new AccumulatingBatchReceiver<>(
+            ApiFutures.<Void>immediateFailedFuture(
+                new IllegalStateException("does nothing, unsuccessfully")));
+    ThresholdBatcher<SimpleBatch> batcher =
+        createSimpleBatcherBuidler(receiver)
+            .setThresholds(BatchingThresholds.<SimpleBatch>create(4))
+            .setFlowController(
+                getTrackedIntegerBatchingFlowController(
+                    3L, null, LimitExceededBehavior.ThrowException))
+            .build();
+
+    assertThat(trackedFlowController.getElementsReserved()).isEqualTo(0);
+    assertThat(trackedFlowController.getElementsReleased()).isEqualTo(0);
+    assertThat(trackedFlowController.getBytesReserved()).isEqualTo(0);
+    assertThat(trackedFlowController.getBytesReleased()).isEqualTo(0);
+
+    batcher.add(SimpleBatch.fromInteger(3));
+    try {
+      batcher.pushCurrentBatch().get();
+      Assert.fail("expected exception");
+    } catch (Exception e) {
+      assertThat(e).isInstanceOf(ExecutionException.class);
```

Comment:
Can we (should we?) interrogate this exception to confirm that it was triggered by the IllegalStateException? e.g. by looking at the inner exception?

Reply:
I think we can do the same by making sure the inner exception is the one we set up. Done.